### PR TITLE
[**Part-1] Add `github` Release Workflow

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Change Log
- Added `release.yml`
- Needs PR #3306

Self automated release for `neo-cli`. Build windows version of `libleveldb` and adds to release of `neo-cli`.

# HOW TO USE
## STEP 1
image

## STEP 2
image

## STEP 3
image

## STEP 4
image

## STEP 5
Wait for job to complete.

---

### Zip/Tarball File Contents
image

### Release Page
image

### Jobs
image

## Type of change

- [x] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Locally and github

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by shargon — DISMISSED** [review]

**Review by Jim8y — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+  cleanup:
+    needs: [build-neo-cli]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Cleanup
+        run: |
+          gh extension install actions/gh-actions-cache
+          
+          echo "Fetching list of cache key"
```
Comment (vncoelho) [inline .github/workflows/release.yml:253]: maybe this can be a script that is called here

**Comment by Jim8y** [comment]
@cschuchardt88 what is the blocker of this pr?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    - name: Lookup Cache Distribution
+      id: cache-leveldb
+      uses: actions/cache@v4
+      with:
+        path: ./leveldb/build/Release/*
+        key: leveldb-windows-${{ matrix.arch }}
+        enableCrossOsArchive: true
+        lookup-only: true
+
+    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
```
Comment (vncoelho) [inline .github/workflows/release.yml:31]: comment all these cache and what they are

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+      name: Cache Distribution
+      uses: actions/cache/save@v4
+      with:
+        path: ./leveldb/build/Release/*
+        key: leveldb-windows-${{ matrix.arch }}
+        enableCrossOsArchive: true
+
+  build-neo-cli:
+    needs: [build-leveldb]
+    name: Build & Publish TarBalls (${{ matrix.runtime }})
```
Comment (vncoelho) [inline .github/workflows/release.yml:69]: Tarballs, all these things

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+      run: |
+        dotnet publish ./src/Neo.CLI \
+        --version-suffix ${{ matrix.runtime }} \
+        --framework net8.0 \
+        --configuration ${{ env.CONFIGURATION }} \
+        --runtime ${{ matrix.runtime }} \
+        --self-contained true \
+        --output ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }} \
+        --verbosity normal \
+        -p:VersionPrefix=${{ env.APP_VERSION }} \
```
Comment (vncoelho) [inline .github/workflows/release.yml:101]: too much parameters as well..

**Review by vncoelho — COMMENTED** [review]
Maybe as readme

File: .github/workflows/release.yml (hunk 1 above)
Comment (vncoelho) [reply .github/workflows/release.yml:253]: @cschuchardt88 , move this Cleanup to a script

File: .github/workflows/release.yml (hunk 4 above)
Comment (Jim8y) [reply .github/workflows/release.yml:101]: Too much parameter? Like those are not necessary?

Comment (vncoelho) [reply .github/workflows/release.yml:101]: I think that a README or comments are necessary, there are several details

**Comment by vncoelho** [comment]
```yml
name: Release (neo-cli)

# Trigger the workflow on a release event when a new release is published
on:
  release:
    types: [published]

# Define environment variables
env:
  DOTNET_VERSION: 8.0.x
  CONFIGURATION: Release
  DIST_PATH: /tmp/dist
  OUTPUT_PATH: /tmp/out

jobs:
  build-leveldb:
    name: Build leveldb (windows-latest)
    runs-on: windows-latest
    strategy:
      matrix:
        arch: [x64, arm64]

    steps:
    # Step to lookup cache for the LevelDB build distribution
    - name: Lookup Cache Distribution
      id: cache-leveldb
      uses: actions/cache@v4
      with:
        path: ./leveldb/build/Release/*
        key: leveldb-windows-${{ matrix.arch }}
        enableCrossOsArchive: true
        lookup-only: true

    # Conditionally checkout LevelDB repository if cache is not found
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Checkout Repository Code (leveldb)
      uses: actions/checkout@v4
      with:
        repository: google/leveldb
        path: leveldb
        submodules: true
        fetch-depth: 0

    # Conditionally setup MSBuild if cache is not found
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Setup MSBuild
      uses: microsoft/setup-msbuild@v2

    # Conditionally setup LevelDB build directory if cache is not found
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Setup LevelDb
      working-directory: ./leveldb
      run: mkdir -p ./build/Release

    # Conditionally create build files for LevelDB if cache is not found
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Create Build Files (win-${{ matrix.arch }})
      working-directory: ./leveldb/build
      run: cmake -DBUILD_SHARED_LIBS=ON -A ${{ matrix.arch }} ..

    # Conditionally build LevelDB using MSBuild if cache is not found
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Build (MSBuild)
      working-directory: ./leveldb/build
      run: msbuild ./leveldb.sln /p:Configuration=Release

    # Conditionally cache the LevelDB distribution if it was built
    - if: ${{ steps.cache-leveldb.outputs.cache-hit != 'true' }}
      name: Cache Distribution
      uses: actions/cache/save@v4
      with:
        path: ./leveldb/build/Release/*
        key: leveldb-windows-${{ matrix.arch }}
        enableCrossOsArchive: true

  build-neo-cli:
    needs: [build-leveldb]
    name: Build & Publish TarBalls (${{ matrix.runtime }})
    runs-on: ubuntu-latest
    strategy:
      matrix:
        runtime: [linux-x64, linux-arm64, win-x64, win-arm64, osx-x64, osx-arm64]

    steps:
    # Step to set the application version from the release tag
    - name: Set Application Version (Environment Variable)
      run: |
        APP_VERSION=$(echo '${{ github.event.release.tag_name }}' | cut -d 'v' -f 2)
        echo "APP_VERSION=$APP_VERSION" >> $GITHUB_ENV

    # Checkout the neo-cli repository code
    - name: Checkout Repository Code
      uses: actions/checkout@v4
      with:
        fetch-depth: 0

    # Setup .NET environment
    - name: Setup .NET
      uses: actions/setup-dotnet@v4
      with:
        dotnet-version: ${{ env.DOTNET_VERSION }}

    # Publish the neo-cli project
    - name: .NET Publish (neo-cli)
      run: |
        dotnet publish ./src/Neo.CLI \
        --version-suffix ${{ matrix.runtime }} \ # Append runtime identifier to version
        --framework net8.0 \ # Target .NET 8.0 framework
        --configuration ${{ env.CONFIGURATION }} \ # Use Release configuration
        --runtime ${{ matrix.runtime }} \ # Specify runtime identifier (e.g., win-x64, linux-arm64)
        --self-contained true \ # Include the .NET runtime with the application
        --output ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }} \ # Output path for the published application
        --verbosity normal \ # Normal verbosity level
        -p:VersionPrefix=${{ env.APP_VERSION }} \ # Set the version prefix to the application version
        -p:RuntimeIdentifier=${{ matrix.runtime }} \ # Set the runtime identifier property
        -p:SelfContained=true \ # Ensure the application is self-contained
        -p:IncludeNativeLibrariesForSelfExtract=false \ # Do not include native libraries for self-extraction
        -p:PublishTrimmed=false \ # Do not trim unused code
        -p:PublishSingleFile=true \ # Publish as a single file
        -p:PublishReadyToRun=true \ # Enable ReadyToRun compilation
        -p:EnableCompressionInSingleFile=true \ # Enable compression in the single file
        -p:DebugType=embedded \ # Embed debug information
        -p:ServerGarbageCollection=true # Enable server garbage collection

    # Build the LevelDBStore plugin
    - name: .NET Build (LevelDBStore)
      run: |
        dotnet build ./src/Plugins/LevelDBStore \
        --version-suffix ${{ matrix.runtime }} \ # Append runtime identifier to version
        --framework net8.0 \ # Target .NET 8.0 framework
        --configuration ${{ env.CONFIGURATION }} \ # Use Release configuration
        --output ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}/Plugins/LevelDBStore \ # Output path for the built plugin
        --verbosity normal \ # Normal verbosity level
        -p:VersionPrefix=${{ env.APP_VERSION }} # Set the version prefix to the application version

    # Remove unnecessary files from the LevelDBStore plugin output
    - name: Remove files (junk)
      working-directory: ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}/Plugins/LevelDBStore
      run: |
        rm -v Neo*
        rm -v *.pdb
        rm -v *.xml

    # Remove XML comment files from the neo-cli output
    - name: Remove Xml Comment Files
      working-directory: ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}
      run: rm -v *.xml

    # Get cached LevelDB distribution for Windows x64 if applicable
    - if: ${{ startsWith(matrix.runtime, 'win-x64') }}
      name: Get Distribution Caches (win-x64)
      uses: actions/cache@v4
      with:
        path: ./leveldb/build/Release/*
        key: leveldb-windows-x64
        enableCrossOsArchive: true
        fail-on-cache-miss: true

    # Get cached LevelDB distribution for Windows arm64 if applicable
    - if: ${{ startsWith(matrix.runtime, 'win-arm64') }}
      name: Get Distribution Caches (win-arm64)
      uses: actions/cache@v4
      with:
        path: ./leveldb/build/Release/*
        key: leveldb-windows-arm64
        enableCrossOsArchive: true
        fail-on-cache-miss: true

    # Copy LevelDB files to the output directory for Windows
    - if: ${{ startsWith(matrix.runtime, 'win') }}
      name: Copy Files (leveldb) (win)
      run: cp -v ./leveldb/build/Release/leveldb.dll ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}/libleveldb.dll

    # Create the distribution directory
    - name: Create Distribution Directory
      run: mkdir -p ${{ env.DIST_PATH }}

    # Create a tarball file for Linux distributions
    - name: Create Tarball File (linux)
      if: ${{ startsWith(matrix.runtime, 'linux') }}
      working-directory: ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}
      run: tar -czvf ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.gz .

    # Create a tarball file for macOS distributions
    - name: Create Tarball File (osx)
      if: ${{ startsWith(matrix.runtime, 'osx') }}
      working-directory: ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}
      run: tar -cJf ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.xz .

    # Create a zip file for Windows distributions
    - name: Create Zip File (win)
      if: ${{ startsWith(matrix.runtime, 'win') }}
      working-directory: ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}
      run: zip ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.zip -r *

    # Create checksum files for Linux distributions
    - name: Create Checksum Files (linux)
      if: ${{ startsWith(matrix.runtime, 'linux') }}
      working-directory: ${{ env.DIST_PATH }}
      env:
        FILENAME: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}
      run: |
          sha256sum ${{ env.FILENAME }}.tar.gz > ${{ env.FILENAME }}.sha256

    # Create checksum files for macOS distributions
    - name: Create Checksum Files (osx)
      if: ${{ startsWith(matrix.runtime, 'osx') }}
      working-directory: ${{ env.DIST_PATH }}
      env:
        FILENAME: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}
      run: |
          sha256sum ${{ env.FILENAME }}.tar.xz > ${{ env.FILENAME }}.sha256

    # Create checksum files for Windows distributions
    - name: Create Checksum Files (win)
      if: ${{ startsWith(matrix.runtime, 'win') }}
      working-directory: ${{ env.DIST_PATH }}
      env:
        FILENAME: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}
      run: |
          sha256sum ${{ env.FILENAME }}.zip > ${{ env.FILENAME }}.sha256

    # List the contents of the distribution and output directories
    - name: Output/Distribution Directory Contents
      run: |
        ls -la ${{ env.DIST_PATH }}
        ls -la ${{ env.OUTPUT_PATH }}/${{ matrix.runtime }}

    # Upload tarball files for Linux distributions
    - name: Upload Tarball File (linux)
      if: ${{ startsWith(matrix.runtime, 'linux') }}
      uses: actions/upload-release-asset@latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
          upload_url: ${{ github.event.release.upload_url }}
          asset_path: ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.gz
          asset_name: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.gz
          asset_content_type: application/x-gtar

    # Upload tarball files for macOS distributions
    - name: Upload Tarball File (osx)
      if: ${{ startsWith(matrix.runtime, 'osx') }}
      uses: actions/upload-release-asset@latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
          upload_url: ${{ github.event.release.upload_url }}
          asset_path: ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.xz
          asset_name: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.tar.xz
          asset_content_type: application/x-gtar

    # Upload zip files for Windows distributions
    - name: Upload Zip File (win)
      if: ${{ startsWith(matrix.runtime, 'win') }}
      uses: actions/upload-release-asset@latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
          upload_url: ${{ github.event.release.upload_url }}
          asset_path: ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.zip
          asset_name: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.zip
          asset_content_type: application/zip

    # Upload checksum files for all distributions
    - name: Upload Checksum File (all)
      uses: actions/upload-release-asset@latest
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
          upload_url: ${{ github.event.release.upload_url }}
          asset_path: ${{ env.DIST_PATH }}/neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.sha256
          asset_name: neo-cli.v${{ env.APP_VERSION }}-${{ matrix.runtime }}.sha256
          asset_content_type: text/plain

  cleanup:
    needs: [build-neo-cli]
    runs-on: ubuntu-latest
    steps:
      # Cleanup step to delete old caches
      - name: Cleanup
        run: |
          gh extension install actions/gh-actions-cache
          
          echo "Fetching list of cache key"
          cacheKeysForPR=$(gh actions-cache list -R $REPO -B $BRANCH -L 100 | cut -f 1 )

          ## Setting this to not fail the workflow while deleting cache keys. 
          set +e
          echo "Deleting caches..."
          for cacheKey in $cacheKeysForPR
          do
              if [ "$cacheKey" != "leveldb-windows-x64" ] && [ "$cacheKey" != "leveldb-windows-arm64" ]
              then
                gh actions-cache delete $cacheKey -R $REPO -B $BRANCH --confirm
              fi
          done
          echo "Done"
        env:
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          REPO: ${{ github.repository }}
          BRANCH: ${{ github.ref }}
```

File: .github/workflows/release.yml (hunk 2 above)
Comment (cschuchardt88) [reply .github/workflows/release.yml:31]: Checks for a cached version of `libleveldb.dll` exists. If doesn't exists than build it

File: .github/workflows/release.yml (hunk 4 above)
Comment (cschuchardt88) [reply .github/workflows/release.yml:101]: These are all the settings i found to work.

**Comment by cschuchardt88** [comment]
@vncoelho Ill be updating it.

**Comment by cschuchardt88** [comment]
@superboyiii can you test it?

**Comment by cschuchardt88** [comment]
This can be merged now. But needs to happen with PR #3338

File: .github/workflows/release.yml (hunk 1 above)
Comment (cschuchardt88) [reply .github/workflows/release.yml:286]: what would the extension be for the file? Type of shell to use? Bash?

**Comment by Jim8y** [comment]
@NGDAdmin

**Comment by superboyiii** [comment]
> @superboyiii can you test it?

I will

**Comment by superboyiii** [comment]
@cschuchardt88

**Comment by cschuchardt88** [comment]
Sorry about my fault, It didn't like the comments in the command-line _(I had too many)_. You found a `dotnet msbuild` bug good job! 😄 Ill report it.

Fix now.

**Comment by superboyiii** [comment]
@cschuchardt88 Yes, the Release action works well now. But do find issue on mac armx64. It's killed at once when execute ./neo-cli

File: .github/workflows/release.yml (hunk 1 above)
Comment (vncoelho) [reply .github/workflows/release.yml:286]: Bourne Again Shell looks good.

**Comment by cschuchardt88** [comment]
> @cschuchardt88 Yes, the Release action works well now. But do find issue on mac armx64. It's killed at once when execute 

Then its must be missing `libleveldb`.

Try `brew install leveldb` command in terminal? If still crashes try `brew info leveldb` command, then _copy_ and _rename_ `leveldb` dll file to `libleveldb` with the right file extension.

**Comment by superboyiii** [comment]
> > @cschuchardt88 Yes, the Release action works well now. But do find issue on mac armx64. It's killed at once when execute
> 
> Then its must be missing `libleveldb`.
> 
> 1st Try allow in settings to allow permission to execute unsigned binary files. `sudo spctl --master-disable`
> 
> image image
> 
> 2nd Try Try `brew install leveldb` command in terminal? If still crashes try `brew info leveldb` command, then _copy_ and _rename_ `leveldb` dll file to `libleveldb` with the right file extension.

Unfortunately，I asked two guys to have a try. They all tried both `brew install leveldb` and adding `libleveldb.dylib` dependency. Either didn't work.

**Comment by cschuchardt88** [comment]
give `codesign --force --deep -s - neo-cli` a try

**Comment by superboyiii** [comment]
> give `codesign --force --deep -s - neo-cli` a try

Works!

**Comment by cschuchardt88** [comment]
> > give `codesign --force --deep -s - neo-cli` a try
> 
> Works!

You retest this PR again. I updated to to run `codesign` command to sign the application. Downloads the `osx` binary and run it to see it crashes again.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,348 @@
+name: Release (neo-cli)
+
+# Trigger the workflow on a release event when a new release is published
+on:
+  release:
+    types: [published]
```
Comment (AnnaShaleva) [inline .github/workflows/release.yml:6]: So how about the manually-triggered builds, don't we need them?

Comment (cschuchardt88) [reply .github/workflows/release.yml:6]: What are you talking about this is automatic on a `publish` release

**Comment by superboyiii** [comment]
> > > give `codesign --force --deep -s - neo-cli` a try
> > 
> > 
> > Works!
> 
> Can you retest this PR again. I updated to run `codesign` command to sign the application. Download the `osx` binary and run it to see it crashes again. But re-run the workflow in this PR, since I updated it.

Sure

**Comment by cschuchardt88** [comment]
This PR resolves issue #3357 on release. But not for locally development.

**Comment by superboyiii** [comment]
@cschuchardt88 Tried osx arm-x64, it works smoothly. Already codesigned.

**Review by superboyiii — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

**Comment by Jim8y** [comment]
LGTM, as long as it works as expected.

**Review by shargon — APPROVED** [review]

**Comment by cschuchardt88** [comment]
> LGTM, since this will only be used by NGD, so as long as it works for ngd as expected.

Doesn't require `dotnet` to be installed also.

**Comment by cschuchardt88** [comment]
> @cschuchardt88 Tried osx arm-x64, it works smoothly. Already codesigned.

Can you test with oracle and other plugins, that require extra `dotnet` libraries to see if it works.

**Comment by superboyiii** [comment]
> > @cschuchardt88 Tried osx arm-x64, it works smoothly. Already codesigned.
> 
> @superboyiii Can you test with oracle and other plugins, that require extra `dotnet` libraries to see if it works.

Works well with Plugins
27d9c42dbfda5d977a0e03168d62e17

**Pull request merged**